### PR TITLE
Workaround for the problem with Generics in SubstMap

### LINKED ISSUE #26: Static function used in trait contract is encoded twice to Viper
The following code snippet produces two identical encodings of `static_pure` in the Viper code. Prusti then panics due to a consistency error in Viper.

## Breaking Code

```rust
extern crate prusti_contracts;

#[pure]
fn static_pure(val: i32) -> i32 {
    val
}

trait Answer {
    #[ensures="result == static_pure(arg)"]
    fn produce_answer(&self, arg: i32) -> i32;
}

struct DeepThought {}

impl Answer for DeepThought {
    fn produce_answer(&self, arg: i32) -> i32 { arg }
}

fn main() {
    let dt = DeepThought {};
    let val1 = static_pure(42);
    let val2 = dt.produce_answer(42);
    assert!(val1 == val2);
}
```

Note that `produce_answer` needs to be defined as a trait function. Implementing it directly on `DeepThought` does not create the erroneous behaviour. Moreover, the `main` function needs to be populated as such; removing any statement other than the `assert` will remove the error.

## Prusti Log

```
  __          __        __  ___             
 |__)  _\/_  |__) |  | /__`  |   ____\/_  |  Hash:  ccf572597ff91ca6c7c
 |      /\   |  \ \__/ .__/  |       /\   |  Build: 2020-04-24 10:57:06

 INFO 2020-04-29T10:21:52Z: prusti::compiler_calls: Parsing of annotations successful (0.0 seconds)
 INFO 2020-04-29T10:21:52Z: prusti::compiler_calls: Type-checking of annotations successful (0.0 seconds)
Verification of 3 items...
 INFO 2020-04-29T10:21:52Z: viper::viper: Using JVM OpenJDK 64-Bit Server VM, Java 11.0.7
 INFO 2020-04-29T10:21:52Z: prusti::verifier: JVM startup (0.3 seconds)
 INFO 2020-04-29T10:21:52Z: viper::verification_context: Using Z3 path: '/usr/bin/viper-z3'
 INFO 2020-04-29T10:21:52Z: viper::verifier: Using backend Silicon version <unknown-build-version>
 INFO 2020-04-29T10:21:53Z: prusti::verifier: Verifier startup (1.62 seconds)
 INFO 2020-04-29T10:21:53Z: prusti_viper::verifier: Received 3 items to be verified:
 INFO 2020-04-29T10:21:53Z: prusti_viper::verifier:  - code::static_pure from code.rs:4:1: 6:2 (code::static_pure[0])
 INFO 2020-04-29T10:21:53Z: prusti_viper::verifier:  - code::main from code.rs:19:1: 24:2 (code::main[0])
 INFO 2020-04-29T10:21:53Z: prusti_viper::verifier:  - <DeepThought as Answer>::produce_answer from code.rs:16:5: 16:54 (code::{{impl}}[0]::produce_answer[0])
 INFO 2020-04-29T10:21:53Z: prusti_viper::encoder::encoder: Encoding: code::static_pure from code.rs:4:1: 6:2 (code::static_pure[0])
 INFO 2020-04-29T10:21:53Z: prusti_viper::encoder::encoder: Encoding: code::main from code.rs:19:1: 24:2 (code::main[0])
 INFO 2020-04-29T10:21:54Z: prusti_viper::encoder::encoder: Encoding: code::static_pure from code.rs:4:1: 6:2 (code::static_pure[0])
 INFO 2020-04-29T10:21:54Z: prusti_viper::encoder::encoder: Encoding: code::static_pure from code.rs:4:1: 6:2 (code::static_pure[0])
 INFO 2020-04-29T10:21:54Z: prusti_viper::encoder::encoder: Encoding: <DeepThought as Answer>::produce_answer from code.rs:16:5: 16:54 (code::{{impl}}[0]::produce_answer[0])
 INFO 2020-04-29T10:21:54Z: prusti_viper::encoder::encoder: Encoding: code::static_pure from code.rs:4:1: 6:2 (code::static_pure[0])
 INFO 2020-04-29T10:21:54Z: prusti_viper::verifier: Encoding to Viper successful (0.15 seconds)
 INFO 2020-04-29T10:21:54Z: prusti_viper::verifier: Viper encoding uses 0 domains, 8 fields, 3 functions, 16 predicates, 5 methods
 INFO 2020-04-29T10:21:54Z: prusti_viper::verifier: Dumping Viper program to '"viper_program"'
 INFO 2020-04-29T10:21:54Z: prusti_viper::verifier: Construction of JVM objects successful (0.40 seconds)
ERROR 2020-04-29T10:21:54Z: viper::verifier: The provided Viper program has 1 consistency errors.
ERROR 2020-04-29T10:21:54Z: viper::verifier: Consistency error: Duplicate identifier m_code$$static_pure$opensqu$0$closesqu$__$TY$__$int$$$int$ found. (<no position>)
thread 'main' panicked at 'Consistency errors. The encoded Viper program is incorrect.', viper/src/verifier.rs:145:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: internal compiler error: unexpected panic

note: the compiler or Prusti unexpectedly panicked.

note: possible reasons include the usage of Rust features that are not supported by Prusti.

note: we would appreciate a bug report for Prusti: <https://github.com/viperproject/prusti-dev/issues>

note: Prusti hash ccf572597ff91ca6c7c built on 2020-04-24 10:57:06

note: rustc <unknown> running on x86_64-unknown-linux-gnu
```

I am running Prusti inside an `ubuntu:bionic` docker container.

### PR DESCRIPTION
**Pull request by JonasAlaif**
A pure function call might not use available Generics resulting in an identical `function_identifier` but differing `key`s. Closes #26

### CODE REVIEW

**Comment by JonasAlaif** [comment]
Closes #26

File: prusti-viper/src/encoder/mir/pure/pure_functions/interface.rs [hunk 1]
```diff
@@ -348,13 +348,20 @@ impl<'v, 'tcx: 'v> PureFunctionEncoderInterface<'tcx>
...
-            };
-            assert!(function_descriptions
-                .insert(function_identifier, description)
-                .is_none());
+            // Then `function_identifier` may be the same with a different `key`.
+            // This is because the substitution map `substs` context may differ,
+            // but the pure function call does not use these generics.
+            // For instance a pure function call in a trait and then a trait impl;
+            // in the former `substs` is empty, but in the latter the generic `Self` is mapped.
+            if function_descriptions.get(&function_identifier).is_none() {
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/pure/pure_functions/interface.rs]:
```suggestion
            if !function_descriptions.contains_key(&function_identifier) {
```

**Review by vakaras — APPROVED** [review]
LGTM

**Review by vakaras — APPROVED** [review]
LGTM

**Comment by vl0w** [comment]
I recently came around an error which looks similar, might it be related?
The error happens in `prusti-viper/src/encoder/encoder.rs:244:9`, see following example.

<details>
<summary>Expand me</summary>

## Error
```
Verification of 7 items...
thread 'rustc' panicked at 'assertion failed: self.functions.borrow_mut().insert(identifier.clone(),\n                                   Rc::new(function))
.is_none()', prusti-viper/src/encoder/encoder.rs:244:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: internal compiler error: unexpected panic
note: Prusti or the compiler unexpectedly panicked. This is a bug.
note: We would appreciate a bug report: https://github.com/viperproject/prusti-dev/issues/new
note: Prusti version: commit 685674eb73 2021-11-25 13:43:07 UTC, built on 2021-12-11 12:57:31 UTC
```

## Code
```rust
extern crate prusti_contracts;
use prusti_contracts::*;

trait MyIterator {
    fn next(&mut self) -> Option<u32>;
}

struct Counter {
    count: u32,
}

#[refine_trait_spec]
impl MyIterator for Counter {
    #[ensures(self.count == old(self.count) + 1)]
    #[ensures(is_some(& result) && unwrap(& result) == old(self.count))]
    fn next(&mut self) -> Option<u32> {
        let res = create_some(self.count);
        self.count += 1;
        return res;
    }
}


fn main() {
    let mut c = Counter { count: 0 };

    let el = c.next();
    assert!(is_some(&el));
    assert!(unwrap(&el) == 0);

    let el = c.next();
    assert!(is_some(&el));
    assert!(unwrap(&el) == 1);

    let el = c.next();
    assert!(is_some(&el));
    assert!(unwrap(&el) == 2);
}


// Option<u32> glue
#[trusted]
#[ensures(is_some(& result))]
#[ensures(unwrap(& result) == val)]
fn create_some(val: u32) -> Option<u32> {
    Some(val)
}

#[trusted]
#[pure]
#[requires(is_some(o))]
fn unwrap(o: &Option<u32>) -> u32 {
    o.unwrap()
}

#[trusted]
#[pure]
#[ensures(matches ! (* o, Some(_)) == result)]
fn is_some(o: &Option<u32>) -> bool {
    o.is_some()
}
```
</details>

**Comment by vakaras** [comment]
> I recently came around an error which looks similar, might it be related? The error happens in `prusti-viper/src/encoder/encoder.rs:244:9`, see following example.

Yes, sounds related.

**Comment by vakaras** [comment]
@JonasAlaif Is the PR ready to be merged?

**Comment by vakaras** [comment]
@JonasAlaif Could you also add a regression test?

**Pull request merged**